### PR TITLE
[FSDP] Fix `use_orig_params=True` + `no_sync()`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90548 [FSDP][Easy] ufmt files
* **#90546 [FSDP] Fix `use_orig_params=True` + `no_sync()`**

`no_sync()` introduces a separate case where a `FlatParameter` maintains an _unsharded_ gradient, instead of a _sharded_ one. This PR fixes `no_sync()` with `use_orig_params=True` by dealing with this separate case.

The existing `use_orig_params=False` already bypasses the built-in parameter/gradient size check, where the `flat_param` is sharded, while the `flat_param.grad` is unsharded. For `use_orig_params=True`, we need to use the same `.data` hack to side step the size check that we used to side step the dtype check for `keep_low_precision_grads=True`.

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -1398,7 +1409,16 @@ def _use_unsharded_grad_views(self) -> None:
...
             )
             param = getattr(module, param_name)
-            param.grad = view
+            if param.shape != view.shape:
+                # NOTE: This is a hack using `.data` to side step the
+                # check that parameter/gradient sizes match. Here,
+                # `param` has the sharded size; `grad` has the unsharded size.
+                # This happens when running in `no_sync()`.
+                if param.grad is None:
+                    param.grad = torch.empty_like(param)
```
Comment (mrshenli) [inline torch/distributed/fsdp/flat_param.py:1418]: I assume this gonna fail if we create a zero-sized tensor here? And this is what you referred to as "the check that parameter/gradient sizes match"?

File: torch/distributed/fsdp/flat_param.py [hunk 2]
```diff
@@ -1398,7 +1409,16 @@ def _use_unsharded_grad_views(self) -> None:
                 f"{self.flat_param._fqns[i]} is missing",
             )
             param = getattr(module, param_name)
-            param.grad = view
+            if param.shape != view.shape:
```
Comment (mrshenli) [inline torch/distributed/fsdp/flat_param.py:1412]: I assume this should only happen under `no_sync` mode. Do we need to assert that?

File: torch/distributed/fsdp/flat_param.py [hunk 3]
```diff
@@ -1413,7 +1433,14 @@ def _use_unsharded_grad_views(self) -> None:
             )  # did not save prefixed name
             param = getattr(module, param_name)
             prim_param = getattr(prim_module, prim_param_name)
-            param.grad = prim_param.grad
+            if param.shape != prim_param.grad.shape:
```
Comment (mrshenli) [inline torch/distributed/fsdp/flat_param.py:1436]: just for my own education, where can I learn more about why do we need to set grad in the `if` clause. I assume we are not going to accumulate grad to sharded param gradients?

File: torch/distributed/fsdp/flat_param.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1412]: Yes, ideally we can assert that, but at the moment, we do not propagate the `state._sync_gradients` attribute to the `FlatParamHandle`. When I wrote this PR a few days ago, I was not confident whether we should do that.

I like the design where the `state` _orchestrates_ the `FlatParamHandle`s. It becomes tricky when we have to maintain coherence between the `state` and `handles`. Currently, we do this for some things like mixed precision, and it causes some problems (as evidenced by my other PR stack).

File: torch/distributed/fsdp/flat_param.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1418]: We should not ever create a zero-sized tensor here because this is `_use_unsharded_grad_views()`, not `_use_sharded_grad_views()`.

The check I am referring to is that for any `param: nn.Parameter`, PyTorch checks that `param` and `param.grad` have the same shape if `param.grad is not None`. However, here, `param` may have the sharded size (e.g. `[5]`), while `param.grad may have the unsharded size (e.g.`[9]`). (The example numbers here show the padding case for 2 ranks and an unsharded size of 10 numel.)

File: torch/distributed/fsdp/flat_param.py (hunk 3 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1436]: For this `if`, the "primary parameter" `prim_param` must have `.grad` set from the previous `for` loop in this method.

Now, this `for` loop is over the shared parameters, and this `if` is just checking the same shape mismatch case as before (i.e. `param` is sharded but its gradient is unsharded).

We should still set the `param.grad` to be `prim_param.grad`, which is a view into the unsharded gradient because for `no_sync()`, we may accumulate an _unsharded_ gradient.

The invariant of `use_orig_params=True` path is that each original parameter and its gradient should always be a view into the underlying `flat_param` and `flat_param.grad`, respectively. Whether the user actually cares that we set `param.grad` to be an unsharded view into `flat_param.grad` during `no_sync()` is debatable, but this is an important design choice that makes reasoning about the system more tractable (at least to me). This is why I was a big proponent of the `_use_sharded_views`, `_use_sharded_grad_views`, `_use_unsharded_views`, `_use_unsharded_grad_views` approach, where any change to the `flat_param` storage should include calls to these methods.

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 4]
```diff
@@ -255,10 +269,12 @@ def _get_subtest_config(self) -> Dict[str, List[Any]]:
             ShardingStrategy.NO_SHARD,
         ],
     )
+    @parametrize("use_orig_params", [True])
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_grad_acc.py:272]: what about use_orig_params=False?

File: test/distributed/fsdp/test_fsdp_use_orig_params.py [hunk 5]
```diff
@@ -1048,10 +1049,122 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        model = nn.Linear(3, 3, bias=False, device="cuda")
+        fsdp_kwargs = {
+            "sharding_strategy": sharding_strategy,
+        }
+        baseline_model = FSDP(
+            copy.deepcopy(model), use_orig_params=False, **fsdp_kwargs
+        )
+        test_model = FSDP(model, use_orig_params=True, **fsdp_kwargs)
+        baseline_optim = torch.optim.AdamW(baseline_model.parameters(), foreach=True)
+        test_optim = torch.optim.AdamW(test_model.parameters(), foreach=True)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_use_orig_params.py:1084]: nit: might be better to just name these model_no_use_orig and model_use_orig so it's clearer which one is which and what we are comparing.

File: test/distributed/fsdp/test_fsdp_use_orig_params.py [hunk 6]
```diff
@@ -1048,10 +1049,122 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+                    test_param.grad,
+                )
+
+        inp = torch.randn((2, 3), device="cuda")
+        grad = torch.rand_like(inp)
+
+        # Compute some reference gradients using one forward/backward
+        baseline_out = baseline_model(inp)
+        test_out = test_model(inp)
+        torch.testing.assert_close(baseline_out, test_out)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_use_orig_params.py:1125]: Is this more of a relaxed assert than `self.assertEqual`? When should we use one over the other?

File: torch/distributed/fsdp/flat_param.py [hunk 7]
```diff
@@ -1247,10 +1247,21 @@ def _use_sharded_flat_param(self) -> None:
         if self._use_orig_params:
             self._use_sharded_views()
             # For the post-forward reshard, we may try to use sharded gradient
-            # views, but for the post-backward reshard, we delay the call to
-            # after the reduce-scatter
+            # views (or unsharded gradient views if a gradient was accumulated
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:1250]: just for my understanding, is this because we want the grad views to be unsharded since we were in no_sync, we didn't actually reduce_scatter and reshard?

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for the thorough analysis and fix!

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 8]
```diff
@@ -128,11 +133,12 @@ def _test_grad_acc(
                 "cpu_offload": cpu_offload,
                 "backward_prefetch": backward_prefetch,
                 "sharding_strategy": sharding_strategy,
+                "use_orig_params": use_orig_params,
             }
             fsdp_model: FSDP = TransformerWithSharedParams.init(
                 self.process_group,
                 FSDPInitMode.RECURSIVE,
-                CUDAInitMode.CUDA_AFTER,
+                CUDAInitMode.CUDA_BEFORE,
```
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py:141]: This was just to silence the warning about having flattening/sharding happening on CPU. There is no reason it needs to be `CUDA_BEFORE`. I am not sure why I did that in the first place.

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 9]
```diff
@@ -197,7 +207,11 @@ def permute_tensor(x: torch.Tensor):
             loss.backward()
             losses.append(loss)
             acc_loss = sum(losses)
-            acc_grads = [p.grad.detach().clone() for p in fsdp_model.parameters()]
+            acc_grads = [
+                p.grad.detach().clone()
+                for p in fsdp_model.parameters()
+                if p.grad is not None
```
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py:213]: `use_orig_params=True` will always require this idiom (for better or for worse) because even though an original parameter requires gradient, it may not receive a gradient if it is not in the local shard. I think this design matches the SPMD spirit best, but it just requires always having this `if grad is not None`.

File: test/distributed/fsdp/test_fsdp_grad_acc.py (hunk 4 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py:272]: Apologies for the window of no signal for this test.

File: test/distributed/fsdp/test_fsdp_use_orig_params.py (hunk 6 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_use_orig_params.py:1125]: Good question. Let me investigate.

File: torch/distributed/fsdp/flat_param.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1250]: Yup

File: test/distributed/fsdp/test_fsdp_use_orig_params.py (hunk 5 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_use_orig_params.py:1084]: Changed.

File: test/distributed/fsdp/test_fsdp_grad_acc.py (hunk 4 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py:272]: Just kidding, the merge failed because this PR is too stale. I will fix in this PR.
